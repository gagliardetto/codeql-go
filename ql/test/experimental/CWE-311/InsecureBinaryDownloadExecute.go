package main

import (
	"io"
	"net/http"
	"os"
	"os/exec"
)

func main() {}

// GOOD
func useHTTPS() {
	remoteURL := "https://example.com/hello.bin"
	localFilepath := "./hello.bin"
	err := downloadFile(localFilepath, remoteURL)
	if err != nil {
		panic(err)
	}

	command := exec.Command("bash", localFilepath)
	_ = command
}

// BAD
func useHTTP() {
	remoteURL := "http://example.com/hello.bin"
	localFilepath := "./hello.bin"
	err := downloadFile(localFilepath, remoteURL)
	if err != nil {
		panic(err)
	}

	command := exec.Command("bash", localFilepath)
	_ = command
}

// GOOD
func useFuncGood() {
	utilDownloadAndExecute("https://example.com/hello.bin")
}

// BAD
func useFuncBad() {
	utilDownloadAndExecute("http://example.com/hello.bin")
}

// BAD
func downloadAndExecuteWithPrefixBad() {
	utilDownloadAndExecuteWithPrefix("./hello.bin")
}

// GOOD
func noExecution() {
	remoteURL := "http://example.com/hello.bin"
	localFilepath := "./hello.bin"
	err := downloadFile(localFilepath, remoteURL)
	if err != nil {
		panic(err)
	}
}

func utilDownloadAndExecute(remoteURL string) {
	localFilepath := "./hello.bin"
	err := downloadFile(localFilepath, remoteURL)
	if err != nil {
		panic(err)
	}

	command := exec.Command("bash", localFilepath)
	_ = command
}
func utilDownloadAndExecuteWithPrefix(name string) {
	remoteURL := "http://example.com/" + name
	localFilepath := "./hello.bin"
	err := downloadFile(localFilepath, remoteURL)
	if err != nil {
		panic(err)
	}

	command := exec.Command("bash", localFilepath)
	_ = command
}

func downloadFile(filepath string, url string) error {
	// Send request:
	resp, err := http.Get(url)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	// Create a new file:
	out, err := os.Create(filepath)
	if err != nil {
		return err
	}
	defer out.Close()

	// Write response body to file:
	_, err = io.Copy(out, resp.Body)
	return err
}
