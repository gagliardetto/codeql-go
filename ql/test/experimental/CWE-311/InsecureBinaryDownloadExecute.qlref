experimental/CWE-311/InsecureBinaryDownloadExecute.ql
