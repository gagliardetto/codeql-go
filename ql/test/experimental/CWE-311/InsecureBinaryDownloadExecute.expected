edges
| InsecureBinaryDownloadExecute.go:27:15:27:44 | "http://example.com/hello.bin" : string | InsecureBinaryDownloadExecute.go:29:37:29:45 | remoteURL : string |
| InsecureBinaryDownloadExecute.go:28:2:28:14 | definition of localFilepath : pointer type | InsecureBinaryDownloadExecute.go:34:34:34:46 | localFilepath |
| InsecureBinaryDownloadExecute.go:29:37:29:45 | remoteURL : string | InsecureBinaryDownloadExecute.go:28:2:28:14 | definition of localFilepath : pointer type |
| InsecureBinaryDownloadExecute.go:45:25:45:54 | "http://example.com/hello.bin" : string | InsecureBinaryDownloadExecute.go:63:29:63:37 | definition of remoteURL : string |
| InsecureBinaryDownloadExecute.go:63:29:63:37 | definition of remoteURL : string | InsecureBinaryDownloadExecute.go:65:37:65:45 | remoteURL : string |
| InsecureBinaryDownloadExecute.go:64:2:64:14 | definition of localFilepath : pointer type | InsecureBinaryDownloadExecute.go:70:34:70:46 | localFilepath |
| InsecureBinaryDownloadExecute.go:65:37:65:45 | remoteURL : string | InsecureBinaryDownloadExecute.go:64:2:64:14 | definition of localFilepath : pointer type |
| InsecureBinaryDownloadExecute.go:74:15:74:35 | "http://example.com/" : string | InsecureBinaryDownloadExecute.go:76:37:76:45 | remoteURL : string |
| InsecureBinaryDownloadExecute.go:75:2:75:14 | definition of localFilepath : pointer type | InsecureBinaryDownloadExecute.go:81:34:81:46 | localFilepath |
| InsecureBinaryDownloadExecute.go:76:37:76:45 | remoteURL : string | InsecureBinaryDownloadExecute.go:75:2:75:14 | definition of localFilepath : pointer type |
nodes
| InsecureBinaryDownloadExecute.go:27:15:27:44 | "http://example.com/hello.bin" : string | semmle.label | "http://example.com/hello.bin" : string |
| InsecureBinaryDownloadExecute.go:28:2:28:14 | definition of localFilepath : pointer type | semmle.label | definition of localFilepath : pointer type |
| InsecureBinaryDownloadExecute.go:29:37:29:45 | remoteURL : string | semmle.label | remoteURL : string |
| InsecureBinaryDownloadExecute.go:34:34:34:46 | localFilepath | semmle.label | localFilepath |
| InsecureBinaryDownloadExecute.go:45:25:45:54 | "http://example.com/hello.bin" : string | semmle.label | "http://example.com/hello.bin" : string |
| InsecureBinaryDownloadExecute.go:63:29:63:37 | definition of remoteURL : string | semmle.label | definition of remoteURL : string |
| InsecureBinaryDownloadExecute.go:64:2:64:14 | definition of localFilepath : pointer type | semmle.label | definition of localFilepath : pointer type |
| InsecureBinaryDownloadExecute.go:65:37:65:45 | remoteURL : string | semmle.label | remoteURL : string |
| InsecureBinaryDownloadExecute.go:70:34:70:46 | localFilepath | semmle.label | localFilepath |
| InsecureBinaryDownloadExecute.go:74:15:74:35 | "http://example.com/" : string | semmle.label | "http://example.com/" : string |
| InsecureBinaryDownloadExecute.go:75:2:75:14 | definition of localFilepath : pointer type | semmle.label | definition of localFilepath : pointer type |
| InsecureBinaryDownloadExecute.go:76:37:76:45 | remoteURL : string | semmle.label | remoteURL : string |
| InsecureBinaryDownloadExecute.go:81:34:81:46 | localFilepath | semmle.label | localFilepath |
#select
| InsecureBinaryDownloadExecute.go:34:34:34:46 | localFilepath | InsecureBinaryDownloadExecute.go:27:15:27:44 | "http://example.com/hello.bin" : string | InsecureBinaryDownloadExecute.go:34:34:34:46 | localFilepath | Executes an insecure download from $@ | InsecureBinaryDownloadExecute.go:27:15:27:44 | "http://example.com/hello.bin" | here |
| InsecureBinaryDownloadExecute.go:70:34:70:46 | localFilepath | InsecureBinaryDownloadExecute.go:45:25:45:54 | "http://example.com/hello.bin" : string | InsecureBinaryDownloadExecute.go:70:34:70:46 | localFilepath | Executes an insecure download from $@ | InsecureBinaryDownloadExecute.go:45:25:45:54 | "http://example.com/hello.bin" | here |
| InsecureBinaryDownloadExecute.go:81:34:81:46 | localFilepath | InsecureBinaryDownloadExecute.go:74:15:74:35 | "http://example.com/" : string | InsecureBinaryDownloadExecute.go:81:34:81:46 | localFilepath | Executes an insecure download from $@ | InsecureBinaryDownloadExecute.go:74:15:74:35 | "http://example.com/" | here |
